roofit/hs3/inc/RooFitHS3/RooJSONFactoryWSTool.h

```diff
@@ -0,0 +1,179 @@
+   static void printImporters();
+   static void printExporters();
+
+   // error handling helpers
+   static void error(const char *s) { throw std::runtime_error(s); }
+   static void error(const std::string &s) { throw std::runtime_error(s); }
+   template <class T>
+   static std::string concat(const T *items, const std::string &sep = ",")
+   {
+      // Returns a string being the concatenation of strings in input list <items>
+      // (names of objects obtained using GetName()) separated by string <sep>.
+      bool first = true;
+      std::string text;
+
+      // iterate over strings in list
+      for (auto it : *items) {
+         if (!first) {
+            // insert separator string
+            text += sep;
+         } else {
+            first = false;
+         }
+         if (!it)
+            text += "NULL";
+         else
+            text += it->GetName();
+      }
+      return text;
+   }
+   template <class T>
+   static std::vector<std::string> names(const T *items)
+   {
+      // Returns a string being the concatenation of strings in input list <items>
+      // (names of objects obtained using GetName()) separated by string <sep>.
+      std::vector<std::string> names;
+      // iterate over strings in list
+      for (auto it : *items) {
+         if (!it)
+            names.push_back("NULL");
+         else
+            names.push_back(it->GetName());
+      }
+      return names;
+   }
+
+   static std::string genPrefix(const RooFit::Detail::JSONNode &p, bool trailing_underscore);
+   static void exportHistogram(const TH1 &h, RooFit::Detail::JSONNode &n, const std::vector<std::string> &obsnames,
+                               const TH1 *errH = 0, bool writeObservables = true, bool writeErrors = true);
+   void exportData(RooAbsData *data, RooFit::Detail::JSONNode &n);
+   static void writeObservables(const TH1 &h, RooFit::Detail::JSONNode &n, const std::vector<std::string> &varnames);
+   static std::vector<std::vector<int>> generateBinIndices(const RooArgList &vars);
+   RooDataHist *
+   readBinnedData(const RooFit::Detail::JSONNode &n, const std::string &namecomp, const RooArgList &observables);
+   static std::map<std::string, RooJSONFactoryWSTool::Var>
+   readObservables(const RooFit::Detail::JSONNode &n, const std::string &obsnamecomp);
+   RooArgSet getObservables(const RooFit::Detail::JSONNode &n, const std::string &obsnamecomp);
+   void setScopeObservables(const RooArgList &args);
+   RooAbsArg *getScopeObject(const std::string &name);
+   void setScopeObject(const std::string &key, RooAbsArg *obj);
+   void clearScope();
+
+   bool importJSON(std::string const& filename);
+   bool importYML(std::string const& filename);
+   bool importJSON(std::istream &os);
+   bool importYML(std::istream &os);
+   bool exportJSON(std::string const& fileName);
+   bool exportYML(std::string const& fileName);
+   bool exportJSON(std::ostream &os);
```

Comment:
Hi, 
 Is this function really needed ? 
 One can always use the export to string and then do:
 ```
 std::ostream out;
out << exportJSONtoString();
```
This will probably avoid the failure in the current test

Reply:
Removing this test function doesn't fix the test failure, it was an internal error in the JSONParser that I now fixed.

I agree that in general the interface needs to be improved, which we can do in a follow-up PR before the release.